File: turtle-soup-bot/src/test/kotlin/io/github/kapu/turtlesoup/utils/AccessControlTest.kt
package io.github.kapu.turtlesoup.utils

import io.github.kapu.turtlesoup.config.AccessConfig
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe

class AccessControlTest : StringSpec({
    "passthrough 모드면 차단 목록도 무시한다" {
        val control = createControl(passthrough = true, blockedUsers = listOf("user-1"))

        control.getDenialReason("user-1", "chat-1") shouldBe null
    }

    "차단된 사용자는 거부한다" {
        val control = createControl(blockedUsers = listOf("user-1"))

        control.getDenialReason("user-1", "chat-1") shouldBe MessageKeys.ERROR_USER_BLOCKED
    }

    "차단된 채팅방은 거부한다" {
        val control = createControl(blockedChats = listOf("chat-1"))

        control.getDenialReason("user-1", "chat-1") shouldBe MessageKeys.ERROR_CHAT_BLOCKED
    }

    "허용 리스트가 설정된 경우 목록 밖 채팅은 거부한다" {
        val control = createControl(allowedChats = listOf("chat-allowed"))

        control.getDenialReason("user-1", "chat-denied") shouldBe MessageKeys.ERROR_ACCESS_DENIED
    }

    "허용 리스트에 있으면 통과한다" {
        val control = createControl(allowedChats = listOf("chat-allowed"))

        control.getDenialReason("user-1", "chat-allowed") shouldBe null
    }
}) {
    companion object {
        private fun createControl(
            passthrough: Boolean = false,
            allowedChats: List<String> = emptyList(),
            blockedChats: List<String> = emptyList(),
            blockedUsers: List<String> = emptyList(),
        ): AccessControl =
            AccessControl(
                AccessConfig(
                    enabled = true,
                    allowedChatIds = allowedChats,
                    blockedChatIds = blockedChats,
                    blockedUserIds = blockedUsers,
                    passthrough = passthrough,
                ),
            )
    }
}

File: game-bot-go/internal/turtlesoup/security/access_control_test.go
package security

import (
	"testing"

	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tsmessages "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/messages"
)

func TestAccessControl_GetDenialReason(t *testing.T) {
	tests := []struct {
		name     string
		cfg      tsconfig.AccessConfig
		userID   string
		chatID   string
		wantDeny bool
		wantMsg  string
	}{
		{
			name:     "Passthrough",
			cfg:      tsconfig.AccessConfig{Passthrough: true},
			userID:   "user1",
			chatID:   "chat1",
			wantDeny: false,
		},
		{
			name:     "Blocked User",
			cfg:      tsconfig.AccessConfig{BlockedUserIDs: []string{"blocked_user"}},
			userID:   "blocked_user",
			chatID:   "chat1",
			wantDeny: true,
			wantMsg:  tsmessages.ErrorUserBlocked,
		},
		{
			name:     "Disabled (implicitly allowing all except blocked users)",
			cfg:      tsconfig.AccessConfig{Enabled: false},
			userID:   "user1",
			chatID:   "chat1",
			wantDeny: false,
		},
		{
			name:     "Blocked Chat",
			cfg:      tsconfig.AccessConfig{Enabled: true, BlockedChatIDs: []string{"blocked_chat"}},
			userID:   "user1",
			chatID:   "blocked_chat",
			wantDeny: true,
			wantMsg:  tsmessages.ErrorChatBlocked,
		},
		{
			name:     "Whitelisted Chat",
			cfg:      tsconfig.AccessConfig{Enabled: true, AllowedChatIDs: []string{"allowed_chat"}},
			userID:   "user1",
			chatID:   "allowed_chat",
			wantDeny: false,
		},
		{
			name:     "Non-Whitelisted Chat",
			cfg:      tsconfig.AccessConfig{Enabled: true, AllowedChatIDs: []string{"allowed_chat"}},
			userID:   "user1",
			chatID:   "other_chat",
			wantDeny: true,
			wantMsg:  tsmessages.ErrorAccessDenied,
		},
		{
			name:     "Empty Whitelist (Allow all if enabled but no whitelist)",
			cfg:      tsconfig.AccessConfig{Enabled: true, AllowedChatIDs: []string{}},
			userID:   "user1",
			chatID:   "chat1",
			wantDeny: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ac := NewAccessControl(tt.cfg)
			got := ac.GetDenialReason(tt.userID, tt.chatID)
			if tt.wantDeny {
				if got == nil {
					t.Errorf("expected denial, got permitted")
				} else if *got != tt.wantMsg {
					t.Errorf("expected denial message %q, got %q", tt.wantMsg, *got)
				}
			} else {
				if got != nil {
					t.Errorf("expected permitted, got denial: %v", *got)
				}
			}
		})
	}
}
